src/FsYacc.Core/fsyaccast.fs

```diff
@@ -58,71 +58,81 @@ let OutputAssoc os p =
         failwithf "%s is given two associativities" key
-    
-    let explicitPrecInfo = 
+
+    let explicitPrecInfo =
         explicitPrecInfo |> Map.ofList
 
     let implicitSymPrecInfo = NoPrecedence
-    let terminals = List.map fst spec.Tokens @ ["error"]
-    let terminalSet = Set.ofList terminals
-    let IsTerminal z = terminalSet.Contains(z)
-    let prec_of_terminal sym implicitPrecInfo = 
+    let terminals = List.map fst spec.Tokens @ [ "error", Range.Empty ]
+    let terminalNamesSet = Set.ofList (terminals |> List.map fst)
+    let IsTerminal z = terminalNamesSet.Contains(z)
+    let prec_of_terminal sym implicitPrecInfo =
        if explicitPrecInfo.ContainsKey(sym) then explicitPrecInfo.[sym]
        else match implicitPrecInfo with Some x -> x | None -> implicitSymPrecInfo
-       
-    let mkSym s = if IsTerminal s then Terminal s else NonTerminal s
-    let prods =  
-        spec.Rules |> List.mapi (fun i (nonterm,rules) -> 
-            rules |> List.mapi (fun j (Rule(syms,precsym,code)) -> 
-                let precInfo = 
-                    let precsym = List.foldBack (fun x acc -> match acc with Some _ -> acc | None -> match x with z when IsTerminal z -> Some z | _ -> acc) syms precsym
+
+    let mkSym ((name, range) as ident) = if IsTerminal name then Terminal ident else NonTerminal ident
+    let prods =
+        spec.Rules |> List.mapi (fun i (nonterm,rules) ->
+            rules |> List.mapi (fun j (Rule(syms,precsym,code)) ->
+                let precInfo =
+                    let precsym =
+                        (syms, precsym)
+                        ||> List.foldBack (fun x acc ->
+                            match acc with
+                            | Some _ -> acc
+                            | None ->
+                                match x with
+                                | (name, _range) as z when IsTerminal name -> Some z
+                                | _ -> acc
+                        )
                     let implicitPrecInfo = NoPrecedence
-                    match precsym with 
-                    | None -> implicitPrecInfo 
+                    match precsym with
+                    | None -> implicitPrecInfo
                     | Some sym -> prec_of_terminal sym None
                 Production(nonterm, precInfo, List.map mkSym syms, code)))
          |> List.concat
     let nonTerminals = List.map fst spec.Rules
-    let nonTerminalSet = Set.ofList nonTerminals
-    let checkNonTerminal nt =  
-        if nt <> "error" && not (nonTerminalSet.Contains(nt)) then 
-            failwith (sprintf "NonTerminal '%s' has no productions" nt)
+    let nonTerminalNamesSet = Set.ofList (nonTerminals |> List.map fst)
+
+    let checkNonTerminal (name, range) =
+        if name <> "error" && not (nonTerminalNamesSet.Contains name) then
+            failwithf "%s(%d,%d): NonTerminal '%s' has no productions" range.startPos.FileName range.startPos.Line range.startPos.Column name
```

Comment:
this is an example of the nicer errors we can get from these structures.  before this is merged I definitely want to convert the entire `.Core` api surface area to return Results containing all collected errors, though. This won't be usable inside an IDE if it throws all over the place.